datafusion/src/physical_plan/hash_join.rs

```diff
@@ -745,9 +747,12 @@ fn build_join_indexes(
                                 &keys_values,
                             )? {
                                 left_indices.append_value(i)?;
-                            } else {
-                                left_indices.append_null()?;
+                                right_indices.append_value(row as u32)?;
```

Comment:
This is the fix, instead of adding the (left=null, right) to every non-match, it does it only once at the end if there wasn't a single match.